### PR TITLE
Layout processinfo with support for custom Format-string

### PR DESCRIPTION
**Pull request by snakefoot**
Makes it possible to make a filename like this:

> fileName="c:/temp/Log/test1-${processinfo:property=StartTime:format=yyyy-MM-dd_HHmmss}.log"

---
This change is [Reviewable](https://reviewable.io/reviews/nlog/nlog/1752)

### CODE REVIEW

File: src/NLog/LayoutRenderers/ProcessInfoLayoutRenderer.cs [hunk 1]
```diff
@@ -81,6 +89,19 @@ protected override void InitializeLayoutRenderer()
             }
 
             this.process = Process.GetCurrentProcess();
+
+            if (!string.IsNullOrEmpty(Format))
```
Comment (304NotModified) [inline src/NLog/LayoutRenderers/ProcessInfoLayoutRenderer.cs]: what about `value.ToStringWithOptionalFormat(...)` ? (in class `FormatHelper`)

**Comment by 304NotModified** [comment]
Good idea! But I think all those reflection isn't needed :) See code comment

**Comment by snakefoot** [comment]
Good observation. I have change the logic to use FormatHelper.ToStringWithOptionalFormat, and it still flies.

File: src/NLog/LayoutRenderers/ProcessInfoLayoutRenderer.cs [hunk 2]
```diff
@@ -107,7 +113,11 @@ protected override void Append(StringBuilder builder, LogEventInfo logEvent)
             if (this.propertyInfo != null)
             {
                 var formatProvider = GetFormatProvider(logEvent);
-                builder.Append(Convert.ToString(this.propertyInfo.GetValue(this.process, null), formatProvider));
+                var value = this.propertyInfo.GetValue(this.process, null);
+                if (value != null && this.Format != null)
+                    builder.Append(Internal.FormatHelper.ToStringWithOptionalFormat(value, this.Format, formatProvider));
```
Comment (304NotModified) [inline src/NLog/LayoutRenderers/ProcessInfoLayoutRenderer.cs]: This method already handles null for both.  ;)

(and it's an extension method)

So my proposal: 

```
 builder.Append(value.ToStringWithOptionalFormat(this.Format, formatProvider));

```

**Comment by snakefoot** [comment]
Yes it looks pretty now :)

**Comment by 304NotModified** [comment]
\0/ 

I the next major version of NLog we should look how generic handle this for all layoutrenderers. In all cases there should be a Format I think.

**Comment by 304NotModified** [comment]
:lgtm:

---

Reviewed 1 of 2 files at r1, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/nlog/nlog/1752#-:-KW5D-eaWonGD6iHuj54:bnfp4nl)*

**Comment by 304NotModified** [comment]
Merged!

**Comment by 304NotModified** [comment]
docs added https://github.com/NLog/NLog/wiki/ProcessInfo-Layout-Renderer

File: src/NLog/LayoutRenderers/ProcessInfoLayoutRenderer.cs [hunk 3]
```diff
@@ -69,6 +69,12 @@ public ProcessInfoLayoutRenderer()
         public ProcessInfoProperty Property { get; set; }
```
Comment (snakefoot) [inline src/NLog/LayoutRenderers/ProcessInfoLayoutRenderer.cs]: @304NotModified Think it should just have DefaultValue, and not DefaultParameter. Make PR to remove DefaultParameter?

Comment (304NotModified) [inline src/NLog/LayoutRenderers/ProcessInfoLayoutRenderer.cs]: yes please :) and add the other as default. Thx in advance!

**Pull request merged**